ludwig/api.py

```diff
@@ -1068,6 +1071,73 @@ def evaluate(
 
             return eval_stats, postproc_predictions, output_directory
 
+    def forecast(
+        self,
+        dataset: DataFrame,
+        data_format: Optional[str] = None,
+        horizon: int = 1,
+        output_directory: Optional[str] = None,
+        output_format: str = "parquet",
+    ) -> DataFrame:
+        # TODO(travis): WIP
+        dataset, _, _, _ = load_dataset_uris(dataset, None, None, None, self.backend)
+        if isinstance(dataset, CacheableDataset):
+            dataset = dataset.unwrap()
+        dataset = load_dataset(dataset, data_format=data_format, df_lib=self.backend.df_engine.df_lib)
+
+        window_sizes = [
+            feature.preprocessing.window_size
+            for feature in self.config_obj.input_features
+            if feature.type == TIMESERIES
+        ]
+        if not window_sizes:
+            raise ValueError("Forecasting requires at least one input feature of type `timeseries`.")
+
+        # TODO(travis): there's a lot of redundancy in this approach, since we are preprocessing the same DataFrame
+        # multiple times with only a small number of features (the horizon) being appended each time.
+        # A much better approach would be to only preprocess a single row, but incorporating the row-level embedding
+        # over the window_size of rows precending it, then performing the model forward pass on only that row of
+        # data.
+        max_lookback_window_size = max(window_sizes)
+        total_forecasted = 0
+        while total_forecasted < horizon:
+            # We only need the last `window_size` worth of rows to forecast the next value
+            dataset = dataset.tail(max_lookback_window_size)
+
+            # Run through preprocessing and prediction to obtain row-wise next values
+            # TODO(travis): can optimize the preprocessing part here, since we only need to preprocess / predict
+            # the last row, not the last `window_size` rows.
+            preds, _ = self.predict(dataset, skip_save_predictions=True, skip_save_unprocessed_output=True)
+
+            next_series = {}
+            for feature in self.config_obj.output_features:
+                if feature.type == TIMESERIES:
+                    key = f"{feature.name}_predictions"
+                    next_series[feature.column] = pd.Series(preds[key].iloc[-1])
+
+            next_preds = pd.DataFrame(next_series)
+            dataset = pd.concat([dataset, next_preds], axis=0).reset_index(drop=True)
```

Comment:
this appraoch is totally fine, but obviously we are preprocessing over and over again datapoints we have already processed, so there's something to be optimzied.

Also from a model perspective, I belive HF does something smart for caching activations for the next steps of text generation, which is kinda similar here potentially, so we can look into that roo.

Reply:
Nice, yeah I think we should definitely think about improvements here. But it definitely depends on how far back / forward we want to forecast. If we're looking at +/- 100 samples this should be fine.